wolfssl/test.h

```diff
@@ -3926,105 +3926,158 @@ static WC_INLINE const char* mymktemp(char *tempfn, int len, int num)
 
 
 
-#if defined(HAVE_SESSION_TICKET) && defined(HAVE_CHACHA) && \
-                                    defined(HAVE_POLY1305)
+#ifdef HAVE_SESSION_TICKET
 
-    #include <wolfssl/wolfcrypt/chacha20_poly1305.h>
+static THREAD_LS_T WC_RNG myKey_rng;
+
+static WC_INLINE void TicketCleanup(void)
+{
+    wc_FreeRng(&myKey_rng);
+}
 
-    typedef struct key_ctx {
-        byte name[WOLFSSL_TICKET_NAME_SZ];        /* name for this context */
-        byte key[CHACHA20_POLY1305_AEAD_KEYSIZE]; /* cipher key */
-    } key_ctx;
+#ifdef OPENSSL_EXTRA
 
-    static THREAD_LS_T key_ctx myKey_ctx;
-    static THREAD_LS_T WC_RNG myKey_rng;
+typedef struct key_ctx {
+    byte name[WOLFSSL_TICKET_NAME_SZ]; /* server name */
+    byte key[AES_256_KEY_SIZE]; /* cipher key */
+    byte hmacKey[WOLFSSL_TICKET_NAME_SZ]; /* hmac key */
+    byte iv[WOLFSSL_TICKET_IV_SZ]; /* cipher iv */
+} key_ctx;
 
-    static WC_INLINE int TicketInit(void)
-    {
-        int ret = wc_InitRng(&myKey_rng);
-        if (ret != 0) return ret;
+static THREAD_LS_T key_ctx myKey_ctx;
 
-        ret = wc_RNG_GenerateBlock(&myKey_rng, myKey_ctx.key, sizeof(myKey_ctx.key));
-        if (ret != 0) return ret;
+static WC_INLINE int TicketInit(void)
```

Comment:
Why are you replacing the wolfSSL ticket encryption?
We don't require or tickets to look like OpenSSL's tickets.

Reply:
That's true. I replaced it only for the OPENSSL_EXTRA layer. Should I make the default internal callback use chacha20-poly1305 and not use OpenSSL API's?

Reply:
The OpenSSL version of the callback needs to call the wolfSSL callback.
If there needs to be a default callback, then yes, use the chacha20-poly1305 version. An AES-GCM version would be good also when ChaCha20-Poly1305 is not available.

Reply:
Changed to chacha-poly and added an 256 bit aes-gcm option.